### PR TITLE
Blazor united initial startup apis for registering pages

### PR DESCRIPTION
**Pull request by javiercn**
Brings in the initial implementation for `Components.Endpoints`.

Note that the goal of this PR is to focus on providing the initial structure and not a proper implementation of any of the concepts. There are links in comments pointing to issues that provide more context and details.

I have stopped at the point where we have a sample that is able to discover components in the app and render a static page. We will be filling in the gaps in comments that are captured in other issues.

### CODE REVIEW

File: src/Components/Endpoints/test/Microsoft.AspNetCore.Components.Endpoints.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (dougbu) [inline src/Components/Endpoints/test/Microsoft.AspNetCore.Components.Endpoints.Tests.csproj:1]: I'm sure this project will be successful across the board 😁:

File: src/Components/Endpoints/src/Microsoft.AspNetCore.Components.Endpoints.csproj [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
+    <Description>Server side rendering for ASP.NET Core Components.</Description>
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Components/Endpoints/src/Microsoft.AspNetCore.Components.Endpoints.csproj:4]: This is following the template from other projects in this folder. It's fine to leave it even if we do not need it.

File: src/Components/Endpoints/src/Microsoft.AspNetCore.Components.Endpoints.csproj [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    <IsAspNetCoreApp>true</IsAspNetCoreApp>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <NoWarn>$(NoWarn);CS0436</NoWarn>
+    <IsPackable>false</IsPackable>
+    <Nullable>annotations</Nullable>
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Components/Endpoints/src/Microsoft.AspNetCore.Components.Endpoints.csproj:10]:
```suggestion
    <Nullable>enable</Nullable>
```

File: src/Components/Endpoints/test/Microsoft.AspNetCore.Components.Endpoints.Tests.csproj (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Endpoints/test/Microsoft.AspNetCore.Components.Endpoints.Tests.csproj:1]: I think he's just remarking that it's a completely empty test project so it's always going to pass :)

Comment (dougbu) [reply src/Components/Endpoints/test/Microsoft.AspNetCore.Components.Endpoints.Tests.csproj:1]: > I think he's just remarking that it's a completely empty test project so it's always going to pass :)

exactly

Comment (javiercn) [reply src/Components/Endpoints/test/Microsoft.AspNetCore.Components.Endpoints.Tests.csproj:1]: It has a couple of tests now 😄.

This PR is only adding all the scaffold for the things we need.

**Comment by javiercn** [comment]
@wtgodbe do you have any idea why the shared framework tests fail?

File: src/Framework/test/TestData.cs [hunk 4]
```diff
@@ -29,6 +29,7 @@ static TestData()
                 "Microsoft.AspNetCore.Authorization.Policy",
                 "Microsoft.AspNetCore.Components",
                 "Microsoft.AspNetCore.Components.Authorization",
+                "Microsoft.AspNetCore.Components.Endpoints",
```
Comment (wtgodbe) [inline src/Framework/test/TestData.cs:32]: @javiercn Add this to `ListedTargetingPackAssemblies` below as well (around line 177)

**Review by SteveSandersonMS — APPROVED** [review]
@javiercn So far this all looks perfect to me.

Do you plan to add an E2E test showing that your new endpoint gets hit? It doesn't necessarily have to use a browser - it could just do an HttpClient request to an E2E test app, like we do for the other prerendering tests. But no worries if you prefer that to be handled in a separate PR.

**Comment by javiercn** [comment]
@SteveSandersonMS let's handle that one separately, as we start getting into more details. Need to check the approach we use in MVC.

**Pull request merged**